ResidualAst breaks on known selectors

**Describe the bug**
`ResidualAst` and `interpreter.PruneAst` don't work correctly when they encounter a selector that it's able to prune. Instead of substituting the entire selector for the constant value, it only substitutes the left-most part, leaving the field selector behind.

**To Reproduce**
Check which components this affects:

- [ ]  parser
- [ ]  checker
- [x]  interpreter

Test setup:
```go
func Test_ResidualAst_Select(t *testing.T) {
	e, _ := cel.NewEnv(
		cel.Declarations(
			decls.NewVar("x", decls.NewMapType(decls.String, decls.Int)),
			decls.NewVar("y", decls.Int),
		),
	)
	ast, _ := e.Parse("x.abc == y")
	prg, _ := e.Program(ast,
		cel.EvalOptions(cel.OptTrackState, cel.OptPartialEval),
	)
	vars, _ := cel.PartialVars(map[string]interface{}{
		"x": map[string]interface{}{"abc": 123},
	}, cel.AttributePattern("y"))
	out, det, err := prg.Eval(vars)
	if !types.IsUnknown(out) {
		t.Fatalf("got %v, expected unknown", out)
	}
	if err != nil {
		t.Fatal(err)
	}
	residual, err := e.ResidualAst(ast, det)
	if err != nil {
		t.Fatal(err)
	}
	expr, err := cel.AstToString(residual)
	if err != nil {
		t.Fatal(err)
	}
	if expr != "123 == y" {
		t.Errorf("got expr: %s, wanted 123 == y", expr)
	}
}
```

**Expected behavior**
This should evaluate the `x.abc` to a constant `123` to be compared to the missing `y`, but instead it produces the garbage `123.abc == y`.

```
=== RUN   Test_ResidualAst_Select
    filter_test.go:87: got expr: 123.abc == y, wanted 123 == y
--- FAIL: Test_ResidualAst_Select (0.00s)
```

**Additional context**
I'm very new to this library, but I believe the problem lies with `interpreter.PruneAst`. In particular perhaps the `*exprpb.Expr_SelectExpr` case in `(*astPruner).prune`. I haven't tried any of the other special cases like calls, lists, etc. so I can't say whether they work correctly or not.

---

**Edit:** As a workaround, select and index expressions can be wrapped in no-op type conversions like `int(x.abc) == y` and they will resolve properly.

@TristonianJones You're dead right about the numeric IDs.

The initial problem is that the various quantifiable attributes (`absoluteAttribute`, `conditionalAttribute`, `maybeAttribute` and `relativeAttribute`) use the unqualified ID and not a more specific ID and `planSelect` defers to them for the ID. So when `evalWatchAttr` calls `state.SetValue` (through `observer`), it's using the ID of `x` rather than of `x.abc`.

That seems to be fixed if I add a `a.id = qual.ID()` to each of the `AddQualifier` methods on those types, though I have no idea whether that is the right fix or not. I'm also not sure if their `Qualify` methods might be broken as well or not as they use the same ID for the qualifiers.

However I've found the same bug with index expressions, and I'm not quite sure how they should be fixed. I'm guessing that `planCallIndex` should persist `expr.ID` through to the returned `Interpretable` somehow. Perhaps the proper fix is for both `planSelect` and `planCallIndex` to properly persist the `expr.ID` by wrapping whatever they would be returning.

I'm really not sure what the correct fix is so I'll leave it up to you, but hopefully all of that helps you with a fix.

Here's an expanded test case with a few more variants of this bug:

```go
func Test_ResidualAst_AttributeQualifiers(t *testing.T) {
	e, _ := cel.NewEnv(
		cel.Declarations(
			decls.NewVar("x", decls.NewMapType(decls.String, decls.Dyn)),
			decls.NewVar("y", decls.NewListType(decls.Int)),
			decls.NewVar("u", decls.Int),
		),
	)
	ast, _ := e.Parse(`x.abc == u && x["abc"] == u && x[x.string] == u && y[0] == u && y[x.zero] == u && (true ? x : y).abc == u && (false ? y : x).abc == u`)
	prg, _ := e.Program(ast,
		cel.EvalOptions(cel.OptTrackState, cel.OptPartialEval),
	)
	vars, _ := cel.PartialVars(map[string]interface{}{
		"x": map[string]interface{}{
			"zero":   0,
			"abc":    123,
			"string": "abc",
		},
		"y": []int{123},
	}, cel.AttributePattern("u"))
	out, det, err := prg.Eval(vars)
	if !types.IsUnknown(out) {
		t.Fatalf("got %v, expected unknown", out)
	}
	if err != nil {
		t.Fatal(err)
	}
	residual, err := e.ResidualAst(ast, det)
	if err != nil {
		t.Fatal(err)
	}
	expr, err := cel.AstToString(residual)
	if err != nil {
		t.Fatal(err)
	}
	const want = "123 == u && 123 == u && 123 == u && 123 == u && 123 == u && 123 == u && 123 == u"
	if expr != want {
		t.Errorf("got expr: %s, wanted %s", expr, want)
	}
}
```

```
=== RUN   Test_ResidualAst_AttributeQualifiers
    filter_test.go:94: got expr: 123.abc == u && 123["abc"] == u && 123[x.string] == u && 123[0] == u && 123[x.zero] == u && 123.abc == u && 123.abc == u, wanted 123 == u && 123 == u && 123 == u && 123 == u && 123 == u && 123 == u && 123 == u
--- FAIL: Test_ResidualAst_AttributeQualifiers (0.00s)
```

--- a/cel/cel_test.go
+++ b/cel/cel_test.go
@@ -1027,3 +1027,44 @@ func Test_Cost(t *testing.T) {
 			min, max, wantedCost)
 	}
 }
+
+func Test_ResidualAst_AttributeQualifiers(t *testing.T) {
+	e, _ := NewEnv(
+		Declarations(
+			decls.NewVar("x", decls.NewMapType(decls.String, decls.Dyn)),
+			decls.NewVar("y", decls.NewListType(decls.Int)),
+			decls.NewVar("u", decls.Int),
+		),
+	)
+	ast, _ := e.Parse(`x.abc == u && x["abc"] == u && x[x.string] == u && y[0] == u && y[x.zero] == u && (true ? x : y).abc == u && (false ? y : x).abc == u`)
+	prg, _ := e.Program(ast,
+		EvalOptions(OptTrackState, OptPartialEval),
+	)
+	vars, _ := PartialVars(map[string]interface{}{
+		"x": map[string]interface{}{
+			"zero":   0,
+			"abc":    123,
+			"string": "abc",
+		},
+		"y": []int{123},
+	}, AttributePattern("u"))
+	out, det, err := prg.Eval(vars)
+	if !types.IsUnknown(out) {
+		t.Fatalf("got %v, expected unknown", out)
+	}
+	if err != nil {
+		t.Fatal(err)
+	}
+	residual, err := e.ResidualAst(ast, det)
+	if err != nil {
+		t.Fatal(err)
+	}
+	expr, err := AstToString(residual)
+	if err != nil {
+		t.Fatal(err)
+	}
+	const want = "123 == u && 123 == u && 123 == u && 123 == u && 123 == u && 123 == u && 123 == u"
+	if expr != want {
+		t.Errorf("got expr: %s, wanted %s", expr, want)
+	}
+}

--- a/interpreter/attributes.go
+++ b/interpreter/attributes.go
@@ -465,7 +465,12 @@ func findMax(x, y int64) int64 {
 //
 // If none of the attributes within the maybe resolves a value, the result is an error.
 func (a *maybeAttribute) AddQualifier(qual Qualifier) (Attribute, error) {
-	str, isStr := qual.(*stringQualifier)
+	str := ""
+	isStr := false
+	cq, isConst := qual.(ConstantQualifier)
+	if isConst {
+		str, isStr = cq.Value().Value().(string)
+	}
 	var augmentedNames []string
 	// First add the qualifier to all existing attributes in the oneof.
 	for _, attr := range a.attrs {
@@ -475,7 +480,7 @@ func (a *maybeAttribute) AddQualifier(qual Qualifier) (Attribute, error) {
 				len(candidateVars),
 				len(candidateVars))
 			for i, name := range candidateVars {
-				augmentedNames[i] = fmt.Sprintf("%s.%s", name, str.value)
+				augmentedNames[i] = fmt.Sprintf("%s.%s", name, str)
 			}
 		}
 		attr.AddQualifier(qual)
@@ -605,7 +610,7 @@ func newQualifier(adapter ref.TypeAdapter, id int64, v interface{}) (Qualifier, 
 	var qual Qualifier
 	switch val := v.(type) {
 	case Attribute:
-		return val, nil
+		return &attrQualifier{id: id, Attribute: val}, nil
 	case string:
 		qual = &stringQualifier{id: id, value: val, celValue: types.String(val), adapter: adapter}
 	case int:
@@ -634,6 +639,20 @@ func newQualifier(adapter ref.TypeAdapter, id int64, v interface{}) (Qualifier, 
 		return nil, fmt.Errorf("invalid qualifier type: %T", v)
 	}
 	return qual, nil
+}
+
+type attrQualifier struct {
+	id int64
+	Attribute
+}
+
+func (q *attrQualifier) ID() int64 {
+	return q.id
+}
+
+// Cost returns zero for constant field qualifiers
+func (q *attrQualifier) Cost() (min, max int64) {
+	return estimateCost(q.Attribute)
 }
 
 type stringQualifier struct {

--- a/interpreter/attributes_test.go
+++ b/interpreter/attributes_test.go
@@ -15,12 +15,18 @@
 package interpreter
 
 import (
+	"reflect"
 	"testing"
 
 	"github.com/golang/protobuf/ptypes"
 
+	"github.com/google/cel-go/checker"
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common"
 	"github.com/google/cel-go/common/containers"
 	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/parser"
 
 	proto3pb "github.com/google/cel-go/test/proto3pb"
 
@@ -508,6 +514,169 @@ func TestAttributes_MissingMsg_UnknownField(t *testing.T) {
 	_, isUnk := out.(types.Unknown)
 	if !isUnk {
 		t.Errorf("got %v, wanted unknown value", out)
+	}
+}
+
+func TestAttribute_StateTracking(t *testing.T) {
+	var tests = []struct {
+		expr  string
+		env   []*exprpb.Decl
+		in    map[string]interface{}
+		out   ref.Val
+		state map[int64]interface{}
+	}{
+		{
+			expr: `[{"field": true}][0].field`,
+			env:  []*exprpb.Decl{},
+			in:   map[string]interface{}{},
+			out:  types.True,
+			state: map[int64]interface{}{
+				// overall expression
+				1: true,
+				// [{"field": true}][0]
+				6: map[ref.Val]ref.Val{types.String("field"): types.True},
+				// [{"field": true}][0].field
+				8: true,
+			},
+		},
+		{
+			expr: `a[1]['two']`,
+			env: []*exprpb.Decl{
+				decls.NewVar("a", decls.NewMapType(
+					decls.Int,
+					decls.NewMapType(decls.String, decls.Bool))),
+			},
+			in: map[string]interface{}{
+				"a": map[int64]interface{}{
+					1: map[string]bool{
+						"two": true,
+					},
+				},
+			},
+			out: types.True,
+			state: map[int64]interface{}{
+				// overall expression
+				1: true,
+				// a[1]
+				2: map[string]bool{"two": true},
+				// a[1]["two"]
+				4: true,
+			},
+		},
+		{
+			expr: `a[1][2][3]`,
+			env: []*exprpb.Decl{
+				decls.NewVar("a", decls.NewMapType(
+					decls.Int,
+					decls.NewMapType(decls.Dyn, decls.Dyn))),
+			},
+			in: map[string]interface{}{
+				"a": map[int64]interface{}{
+					1: map[int64]interface{}{
+						1: 0,
+						2: []string{"index", "middex", "outdex", "dex"},
+					},
+				},
+			},
+			out: types.String("dex"),
+			state: map[int64]interface{}{
+				// overall expression
+				1: "dex",
+				// a[1]
+				2: map[int64]interface{}{
+					1: 0,
+					2: []string{"index", "middex", "outdex", "dex"},
+				},
+				// a[1][2]
+				4: []string{"index", "middex", "outdex", "dex"},
+				// a[1][2][3]
+				6: "dex",
+			},
+		},
+		{
+			expr: `a[1][2][a[1][1]]`,
+			env: []*exprpb.Decl{
+				decls.NewVar("a", decls.NewMapType(
+					decls.Int,
+					decls.NewMapType(decls.Dyn, decls.Dyn))),
+			},
+			in: map[string]interface{}{
+				"a": map[int64]interface{}{
+					1: map[int64]interface{}{
+						1: 0,
+						2: []string{"index", "middex", "outdex", "dex"},
+					},
+				},
+			},
+			out: types.String("index"),
+			state: map[int64]interface{}{
+				// overall expression
+				1: "index",
+				// a[1]
+				2: map[int64]interface{}{
+					1: 0,
+					2: []string{"index", "middex", "outdex", "dex"},
+				},
+				// a[1][2]
+				4: []string{"index", "middex", "outdex", "dex"},
+				// a[1][2][a[1][1]]
+				6: "index",
+				// dynamic index into a[1][2]
+				// a[1]
+				8: map[int64]interface{}{
+					1: 0,
+					2: []string{"index", "middex", "outdex", "dex"},
+				},
+				// a[1][1]
+				10: int64(0),
+			},
+		},
+	}
+	for _, test := range tests {
+		tc := test
+		t.Run(tc.expr, func(tt *testing.T) {
+			src := common.NewTextSource(tc.expr)
+			parsed, errors := parser.Parse(src)
+			if len(errors.GetErrors()) != 0 {
+				tt.Fatalf(errors.ToDisplayString())
+			}
+			cont := containers.DefaultContainer
+			reg := types.NewRegistry()
+			env := checker.NewStandardEnv(cont, reg)
+			if tc.env != nil {
+				env.Add(tc.env...)
+			}
+			checked, errors := checker.Check(parsed, src, env)
+			if len(errors.GetErrors()) != 0 {
+				tt.Fatalf(errors.ToDisplayString())
+			}
+			attrs := NewAttributeFactory(cont, reg, reg)
+			interp := NewStandardInterpreter(cont, reg, reg, attrs)
+			// Show that program planning will now produce an error.
+			st := NewEvalState()
+			i, err := interp.NewInterpretable(checked, Optimize(), TrackState(st))
+			if err != nil {
+				tt.Fatal(err)
+			}
+			if err != nil {
+				tt.Fatal(err)
+			}
+			in, _ := NewActivation(tc.in)
+			out := i.Eval(in)
+			if tc.out.Equal(out) != types.True {
+				tt.Errorf("got %v, wanted %v", out.Value(), tc.out)
+			}
+			for id, val := range tc.state {
+				stVal, found := st.Value(id)
+				if !found {
+					tt.Errorf("state not found for %d=%v", id, val)
+					continue
+				}
+				if !reflect.DeepEqual(stVal.Value(), val) {
+					tt.Errorf("got %v, wanted %v for id: %d", stVal.Value(), val, id)
+				}
+			}
+		})
 	}
 }
 

--- a/interpreter/interpretable.go
+++ b/interpreter/interpretable.go
@@ -60,7 +60,14 @@ type InterpretableAttribute interface {
 	// Attribute, the Attribute should first be copied before adding the qualifier. Attributes
 	// are not copyable by default, so this is a capable that would need to be added to the
 	// AttributeFactory or specifically to the underlying Attribute implementation.
-	AddQualifier(Qualifier) (InterpretableAttribute, error)
+	AddQualifier(Qualifier) (Attribute, error)
+
+	// Qualify replicates the Attribute.Qualify method to permit extension and interception
+	// of object qualification.
+	Qualify(vars Activation, obj interface{}) (interface{}, error)
+
+	// Resolve returns the value of the Attribute given the current Activation.
+	Resolve(Activation) (interface{}, error)
 }
 
 // InterpretableCall interface for inspecting Interpretable instructions related to function calls.
@@ -101,7 +108,7 @@ func (test *evalTestOnly) Eval(ctx Activation) ref.Val {
 	if test.fieldType != nil {
 		opAttr, ok := test.op.(InterpretableAttribute)
 		if ok {
-			opVal, err := opAttr.Attr().Resolve(ctx)
+			opVal, err := opAttr.Resolve(ctx)
 			if err != nil {
 				return types.NewErr(err.Error())
 			}
@@ -822,16 +829,94 @@ type evalWatchAttr struct {
 	observer evalObserver
 }
 
-// Eval implements the Interpretable interface method.
-func (e *evalWatchAttr) Eval(vars Activation) ref.Val {
-	val := e.InterpretableAttribute.Eval(vars)
+// AddQualifier creates a wrapper over the incoming qualifier which observes the qualification
+// result.
+func (e *evalWatchAttr) AddQualifier(q Qualifier) (Attribute, error) {
+	cq, isConst := q.(ConstantQualifier)
+	if isConst {
+		q = &evalWatchConstQual{
+			ConstantQualifier: cq,
+			observer:          e.observer,
+			adapter:           e.InterpretableAttribute.Adapter(),
+		}
+	} else {
+		q = &evalWatchQual{
+			Qualifier: q,
+			observer:  e.observer,
+			adapter:   e.InterpretableAttribute.Adapter(),
+		}
+	}
+	_, err := e.InterpretableAttribute.AddQualifier(q)
+	return e, err
+}
+
+// evalWatchConstQual observes the qualification of an object using a constant boolean, int,
+// string, or uint.
+type evalWatchConstQual struct {
+	ConstantQualifier
+	observer evalObserver
+	adapter  ref.TypeAdapter
+}
+
+// Cost implements the Coster interface method.
+func (e *evalWatchConstQual) Cost() (min, max int64) {
+	return estimateCost(e.ConstantQualifier)
+}
+
+// Qualify observes the qualification of a object via a constant boolean, int, string, or uint.
+func (e *evalWatchConstQual) Qualify(vars Activation, obj interface{}) (interface{}, error) {
+	out, err := e.ConstantQualifier.Qualify(vars, obj)
+	var val ref.Val
+	if err != nil {
+		val = types.NewErr(err.Error())
+	} else {
+		val = e.adapter.NativeToValue(out)
+	}
 	e.observer(e.ID(), val)
-	return val
+	return out, err
+}
+
+// QualifierValueEquals tests whether the incoming value is equal to the qualificying constant.
+func (e *evalWatchConstQual) QualifierValueEquals(value interface{}) bool {
+	qve, ok := e.ConstantQualifier.(qualifierValueEquator)
+	return ok && qve.QualifierValueEquals(value)
+}
+
+// evalWatchQual observes the qualification of an object by a value computed at runtime.
+type evalWatchQual struct {
+	Qualifier
+	observer evalObserver
+	adapter  ref.TypeAdapter
+}
+
+// Cost implements the Coster interface method.
+func (e *evalWatchQual) Cost() (min, max int64) {
+	return estimateCost(e.Qualifier)
+}
+
+// Qualify observes the qualification of a object via a value computed at runtime.
+func (e *evalWatchQual) Qualify(vars Activation, obj interface{}) (interface{}, error) {
+	out, err := e.Qualifier.Qualify(vars, obj)
+	var val ref.Val
+	if err != nil {
+		val = types.NewErr(err.Error())
+	} else {
+		val = e.adapter.NativeToValue(out)
+	}
+	e.observer(e.ID(), val)
+	return out, err
 }
 
 // Cost implements the Coster interface method.
 func (e *evalWatchAttr) Cost() (min, max int64) {
 	return estimateCost(e.InterpretableAttribute)
+}
+
+// Eval implements the Interpretable interface method.
+func (e *evalWatchAttr) Eval(vars Activation) ref.Val {
+	val := e.InterpretableAttribute.Eval(vars)
+	e.observer(e.ID(), val)
+	return val
 }
 
 // evalWatchConst describes a watcher of an instConst Interpretable.
@@ -1066,6 +1151,7 @@ func (fold *evalExhaustiveFold) Cost() (min, max int64) {
 		iMax + aMax + cMax*rangeCnt + sMax*rangeCnt + rMax
 }
 
+// evalAttr evaluates an Attribute value.
 type evalAttr struct {
 	adapter ref.TypeAdapter
 	attr    Attribute
@@ -1074,6 +1160,13 @@ type evalAttr struct {
 // ID of the attribute instruction.
 func (a *evalAttr) ID() int64 {
 	return a.attr.ID()
+}
+
+// AddQualifier implements the instAttr interface method.
+func (a *evalAttr) AddQualifier(qual Qualifier) (Attribute, error) {
+	attr, err := a.attr.AddQualifier(qual)
+	a.attr = attr
+	return attr, err
 }
 
 // Attr implements the instAttr interface method.
@@ -1086,6 +1179,11 @@ func (a *evalAttr) Adapter() ref.TypeAdapter {
 	return a.adapter
 }
 
+// Cost implements the Coster interface method.
+func (a *evalAttr) Cost() (min, max int64) {
+	return estimateCost(a.attr)
+}
+
 // Eval implements the Interpretable interface method.
 func (a *evalAttr) Eval(ctx Activation) ref.Val {
 	v, err := a.attr.Resolve(ctx)
@@ -1095,13 +1193,12 @@ func (a *evalAttr) Eval(ctx Activation) ref.Val {
 	return a.adapter.NativeToValue(v)
 }
 
-// Cost implements the Coster interface method.
-func (a *evalAttr) Cost() (min, max int64) {
-	return estimateCost(a.attr)
+// Qualify proxies to the Attribute's Qualify method.
+func (a *evalAttr) Qualify(ctx Activation, obj interface{}) (interface{}, error) {
+	return a.attr.Qualify(ctx, obj)
 }
 
-// AddQualifier implements the instAttr interface method.
-func (a *evalAttr) AddQualifier(qual Qualifier) (InterpretableAttribute, error) {
-	_, err := a.attr.AddQualifier(qual)
-	return a, err
+// Resolve proxies to the Attribute's Resolve method.
+func (a *evalAttr) Resolve(ctx Activation) (interface{}, error) {
+	return a.attr.Resolve(ctx)
 }

--- a/interpreter/interpreter_test.go
+++ b/interpreter/interpreter_test.go
@@ -1142,7 +1142,7 @@ func TestInterpreter(t *testing.T) {
 			if tc.cost != nil {
 				minCost, maxCost := estimateCost(prg)
 				if minCost != tc.cost[0] || maxCost != tc.cost[1] {
-					t.Errorf("Got cost interval [%v, %v], wanted %v", minCost, maxCost, tc.cost)
+					tt.Errorf("Got cost interval [%v, %v], wanted %v", minCost, maxCost, tc.cost)
 				}
 			}
 			state := NewEvalState()
@@ -1177,8 +1177,8 @@ func TestInterpreter(t *testing.T) {
 						}
 						minCost, maxCost := estimateCost(prg)
 						if minCost != wantedCost[0] || maxCost != wantedCost[1] {
-							t.Errorf("Got exhaustive cost interval [%v, %v], wanted %v",
-								minCost, maxCost, tc.cost)
+							ttt.Errorf("Got exhaustive cost interval [%v, %v], wanted %v",
+								minCost, maxCost, wantedCost)
 						}
 					}
 					if mode == "optimize" && tc.cost != nil {
@@ -1188,7 +1188,7 @@ func TestInterpreter(t *testing.T) {
 						}
 						minCost, maxCost := estimateCost(prg)
 						if minCost != wantedCost[0] || maxCost != wantedCost[1] {
-							t.Errorf("Got optimize cost interval [%v, %v], wanted %v", minCost, maxCost, tc.cost)
+							ttt.Errorf("Got optimize cost interval [%v, %v], wanted %v", minCost, maxCost, tc.cost)
 						}
 					}
 					state.Reset()

--- a/interpreter/planner.go
+++ b/interpreter/planner.go
@@ -228,7 +228,8 @@ func (p *planner) planSelect(expr *exprpb.Expr) (Interpretable, error) {
 	// Lastly, create a field selection Interpretable.
 	attr, isAttr := op.(InterpretableAttribute)
 	if isAttr {
-		return attr.AddQualifier(qual)
+		_, err = attr.AddQualifier(qual)
+		return attr, err
 	}
 
 	relAttr := p.attrFactory.RelativeAttribute(op.ID(), op)
@@ -480,24 +481,41 @@ func (p *planner) planCallIndex(expr *exprpb.Expr,
 	if !isOpAttr {
 		opAttr = &evalAttr{
 			adapter: p.adapter,
-			attr:    p.attrFactory.RelativeAttribute(expr.Id, op),
+			attr:    p.attrFactory.RelativeAttribute(op.ID(), op),
 		}
 	}
+	decAttr, err := p.decorate(opAttr, nil)
+	if err != nil {
+		return nil, err
+	}
+	opAttr, isOpAttr = decAttr.(InterpretableAttribute)
+	if !isOpAttr {
+		return nil, fmt.Errorf("invalid attribute decoration: %v(%T)", decAttr, decAttr)
+	}
+	opType := p.typeMap[expr.GetCallExpr().GetTarget().GetId()]
 	indConst, isIndConst := ind.(InterpretableConst)
 	if isIndConst {
-		opType := p.typeMap[expr.GetCallExpr().GetTarget().GetId()]
 		qual, err := p.attrFactory.NewQualifier(
-			opType, indConst.ID(), indConst.Value())
+			opType, expr.GetId(), indConst.Value())
 		if err != nil {
 			return nil, err
 		}
-		return opAttr.AddQualifier(qual)
+		_, err = opAttr.AddQualifier(qual)
+		return opAttr, err
 	}
 	indAttr, isIndAttr := ind.(InterpretableAttribute)
 	if isIndAttr {
-		return opAttr.AddQualifier(indAttr.Attr())
+		qual, err := p.attrFactory.NewQualifier(
+			opType, expr.GetId(), indAttr)
+		if err != nil {
+			return nil, err
+		}
+		_, err = opAttr.AddQualifier(qual)
+		return opAttr, err
 	}
-	return opAttr.AddQualifier(p.attrFactory.RelativeAttribute(ind.ID(), ind))
+	_, err = opAttr.AddQualifier(
+		p.attrFactory.RelativeAttribute(expr.GetId(), ind))
+	return opAttr, err
 }
 
 // planCreateList generates a list construction Interpretable.

